lec07-pop-models.Rmd

```diff
@@ -1,4 +1,617 @@
+How small is small, and why does it have to be small? The answers to these are related, and to see why,
+let's look at what can happen if $dt$ is too big with an example from the logistic equation. 
+
+```{r}
+# what happens if we make dt too large when simulating? 
+# in this example, the starting population N0 is much larger than the carrying capacity.
+# This means dN/dt will be a large negative number: N will decrease quickly from such a large size.
+
+r <- 0.5
+K <- 50
+dt <- 1
+N0 <- 400
+
+dN <- logistic_eqn(N0, r, K)*dt # this is the change in N that we wil add to N0 to calculate N1
+
+N1 <- N0 + dN
+
+print(N1) # this is no good: N has shot from +400 to -1000 in a single step! 
+
+dN <- logistic_eqn(N1, r, K)*dt # calculate what the next change in N would be
+
+N2 <- N1 + dN
+
+print(N2) # N is getting more and more negative!
+ 
+```
+
+### Challenge - in groups
+
+Use Euler's method to calculate a numerical solution for the following differential equation:
+
+$$\frac{dx}{dt} = -2.3 x$$
+
+  - Try a few different step sizes ($\Delta t$) such as $0.7$ and $1$. What happens if $\Delta t$ is too large?
+
+This is what's called **numerical instability** --- the numerical solution grows very large 
+(either positive or negative) but the true solution doesn't. In the logistic equation example above,
+this numerical instability can happen if $N$ falls below $0$: the population will keep getting
+more and more negative after that. 
+
+## Using `deSolve` in R to solve differential equations numerically
+
+R has a package called `deSolve` which contains several functions to numerically solve differential equations.
+Let's look at the documentation for `deSolve`:
+
+```{r}
+library(deSolve)
+help(deSolve)
+```
+
+This is pretty hard to read, especially if your calculus was a long time ago. The important thing to know for
+our purposes is that the equations we're working with are **first-order ordinary differential equations**.
+*First-order* means that the derivatives are all first derivatives (as opposed to second derivatives, third
+derivatives, etc.). **Ordinary** means that the derivatives are full derivatives (as opposed to partial
+derivatives which are written $\partial N / \partial t$). 
+
+This means that the function we'll be using is called `ode`, which stands for 'ordinary differential equation'. 
+
+```{r}
+help(ode)
+# The possible methods are all different ways of numerically solving DEs.
+
+#stiff vs. non-stiff : no precise definition, but 'stiff' generally means it's possible to be numerically unstable if the step size is too big. 
+```
+
+The method used by `ode` to numerically solve differential equations is called `lsoda`, which automatically
+tries to use the best method depending on the equation. This is why people like to use functions like `ode`:
+the result is often more accurate than Euler's method and might also be faster to run. 
+
+```{r}
+# Example: deSolve with logistic eqn
+
+library(deSolve)
```

Comment:
You already did this above

Reply:
oookkaaaayy

whatev

why can't I import things a lot